### PR TITLE
[Model] Add Wan2.2 text-to-video support

### PR DESCRIPTION
**Pull request by linyueqian**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Add support for Wan2.2 text-to-video generation.

## Test Plan
  ```bash
  python examples/offline_inference/wan22/text_to_video.py \
      --prompt "Two anthropomorphic cats in comfy boxing gear and bright gloves fight intensely on a spotlighted stage." \
      --negative_prompt "色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，
  JPEG压缩残留，丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止不
  动的画面，杂乱的背景，三条腿，背景人很多，倒着走" \
      --height 720 \
      --width 1280 \
      --num_frames 32 \
      --guidance_scale 4.0 \
      --guidance_scale_high 3.0 \
      --num_inference_steps 40 \
      --fps 16 \
      --output t2v_out.mp4

```
## Test Result

https://github.com/user-attachments/assets/822fd347-b8fe-4f03-b588-ac271c83e18b

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/omni_diffusion.py [hunk 1]
```diff
@@ -52,7 +52,10 @@ def __init__(self, od_config: OmniDiffusionConfig | None = None, **kwargs):
             "model_index.json",
             od_config.model,
         )
-        od_config.model_class_name = config_dict.get("_class_name", None)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/omni_diffusion.py:55]: why change this? it seems not necessary

File: vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py [hunk 2]
```diff
@@ -0,0 +1,851 @@
...
+        if missing_keys:
+            logger.warning(f"Weights not found in model ({len(missing_keys)} keys): {missing_keys[:10]}...")
+
+        # Check for uninitialized params
+        uninitialized = [k for k in params_dict.keys() if k not in loaded_keys]
+        if uninitialized:
+            logger.warning(f"Model params not in checkpoint ({len(uninitialized)} keys): {uninitialized[:10]}...")
+
+    @classmethod
+    def from_pretrained(
```
Comment (ZJY0516) [inline vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py]: You can refer to the `load_weights` function in https://github.com/vllm-project/vllm-omni/pull/157

File: vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py [hunk 3]
```diff
@@ -0,0 +1,851 @@
...
+                for _ in range(num_layers)
+            ]
+        )
+
+        # 4. Output norm & projection
+        self.norm_out = FP32LayerNorm(inner_dim, eps, elementwise_affine=False)
+        self.proj_out = nn.Linear(inner_dim, out_channels * math.prod(patch_size))
+        self.scale_shift_table = nn.Parameter(torch.randn(1, 2, inner_dim) / inner_dim**0.5)
+
+        self.gradient_checkpointing = False
```
Comment (ZJY0516) [inline vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py]: vllm-omni does not need this

File: vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2.py [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+        )
+        latents_std = 1.0 / torch.tensor(self.vae.config.latents_std).view(1, self.vae.config.z_dim, 1, 1, 1).to(
+            latents.device, latents.dtype
+        )
+        latents = latents / latents_std + latents_mean
+        decoded = self.vae.decode(latents, return_dict=False)[0]
+        if output_type == "latent":
+            output = latents
+        else:
+            output = self.video_processor.postprocess_video(decoded, output_type=output_type or "np")
```
Comment (ZJY0516) [inline vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2.py]: This should be put in `post_process_func` function

**Comment by SamitHuang** [comment]
nice work. Can you try to increase `--num_inference_steps=10` to like 100 and check whether the video quality become normal?

File: vllm_omni/diffusion/omni_diffusion.py (hunk 1 above)
Comment (linyueqian) [reply vllm_omni/diffusion/omni_diffusion.py:55]: i will revert it.

File: vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py (hunk 2 above)
Comment (linyueqian) [reply vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py]: i saw that it has been merged, i will test if i can use similar loading logic.

File: vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py (hunk 3 above)
Comment (linyueqian) [reply vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py]: the gradient_checkpointing has been deleted.

File: vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2.py (hunk 4 above)
Comment (linyueqian) [reply vllm_omni/diffusion/models/wan2_2/pipeline_wan2_2.py]: done.

**Comment by linyueqian** [comment]
> nice work. Can you try to increase `--num_inference_steps=10` to like 100 and check whether the video quality become normal?

@SamitHuang I try with 40 steps and it takes about five minutes to generate.

https://github.com/user-attachments/assets/f74c671e-71a5-4266-9316-77a868d05c96

**Comment by hsliuustc0106** [comment]
> > nice work. Can you try to increase `--num_inference_steps=10` to like 100 and check whether the video quality become normal?
> 
> @SamitHuang I try with 40 steps and it takes about five minutes to generate.
> 
>  wan22_output_50.mp4

we may need some acceleration methods to speed up generation

**Comment by hsliuustc0106** [comment]
please add this model to the supported models.md

File: examples/offline_inference/wan22/text_to_video.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    else:
+        video_array = frames
+        if hasattr(video_array, "shape") and video_array.ndim == 5:
+            video_array = video_array[0]
+
+    # Convert 4D array (frames, H, W, C) to list of frames for export_to_video
+    if isinstance(video_array, np.ndarray) and video_array.ndim == 4:
+        video_array = list(video_array)
+
+    export_to_video(video_array, str(output_path), fps=16)
```
Comment (SamitHuang) [inline examples/offline_inference/wan22/text_to_video.py]: fps can be 24 too. it's better to be configurable via argparser

Comment (linyueqian) [reply examples/offline_inference/wan22/text_to_video.py]: got it. i change it accordingly

**Comment by SamitHuang** [comment]
i think 

> > nice work. Can you try to increase `--num_inference_steps=10` to like 100 and check whether the video quality become normal?
> 
> @SamitHuang I try with 40 steps and it takes about five minutes to generate.
> 
>  wan22_output_50.mp4

i think you can update the test method and result with this new video, where `num_frames` seems increased. btw, diffusers example applies `negative_prompt`, we should apply it in the test as well to verify CFG works

**Comment by hsliuustc0106** [comment]
image
this PR only supports t2v, right?

**Comment by linyueqian** [comment]
> image this PR only supports t2v, right?

yes.

**Comment by linyueqian** [comment]
> i think
> 
> > > nice work. Can you try to increase `--num_inference_steps=10` to like 100 and check whether the video quality become normal?
> > 
> > 
> > @SamitHuang I try with 40 steps and it takes about five minutes to generate.
> > wan22_output_50.mp4
> 
> i think you can update the test method and result with this new video, where `num_frames` seems increased. btw, diffusers example applies `negative_prompt`, we should apply it in the test as well to verify CFG works

I have updated the test result in the first comment.

**Comment by hufangjian** [comment]
when diffuser model support TP,CFG,USP and distVAE?

File: vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py [hunk 6]
```diff
@@ -0,0 +1,797 @@
...
+        hidden_states_img = None
+        if encoder_hidden_states_img is not None:
+            key_img, _ = self.add_k_proj(encoder_hidden_states_img)
+            value_img, _ = self.add_v_proj(encoder_hidden_states_img)
+            key_img = self.norm_added_k(key_img)
+
+            key_img = key_img.unflatten(2, (self.num_heads, -1))
+            value_img = value_img.unflatten(2, (self.num_heads, -1))
+
+            hidden_states_img = dispatch_attention_fn(
```
Comment (ZJY0516) [inline vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py]: We have had a unified attention layer. Could you please try to use that?

File: vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py [hunk 7]
```diff
@@ -0,0 +1,797 @@
...
+            else:
+                param = params_dict[name]
+                weight_loader = getattr(param, "weight_loader", default_weight_loader)
+                weight_loader(param, loaded_weight)
+            loaded_params.add(name)
+
+        return loaded_params
+
+    @classmethod
+    def from_pretrained(
```
Comment (ZJY0516) [inline vllm_omni/diffusion/models/wan2_2/wan2_2_transformer.py]: Don't use `from_pretrained` for initialization to align with other model definitions

**Comment by hsliuustc0106** [comment]
> when diffuser model support TP,CFG,USP and distVAE?

TP/USP should be ready by the end of this month, others left to Q1

**Comment by hsliuustc0106** [comment]
add the tests, please refer to the qwen-image tests

**Comment by linyueqian** [comment]
> add the tests, please refer to the qwen-image tests

got it. i just add the `test_video_diffusion_model.py` file in a similar fashion.

File: tests/single_stage/test_video_diffusion_model.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import os
```
Comment (hsliuustc0106) [inline tests/single_stage/test_video_diffusion_model.py:1]: @SamitHuang please check whether WAN2.2 can run on L4 GPU

File: tests/single_stage/test_video_diffusion_model.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        width=width,
+        num_frames=num_frames,
+        num_inference_steps=2,
+        guidance_scale=1.0,
+        generator=torch.Generator("cuda").manual_seed(42),
+    )
+
+    assert frames is not None
+    assert hasattr(frames, "shape")
+    assert frames.shape[1] == height
```
Comment (hsliuustc0106) [inline tests/single_stage/test_video_diffusion_model.py]: do we need to check frames.shape[0]?

Comment (linyueqian) [reply tests/single_stage/test_video_diffusion_model.py]: number of frames may not be exactly the same as input. so i change the num_frames in test.

**Comment by hsliuustc0106** [comment]
I think we can get this PR merged now, later we need to open a new issue for a few todo jobs

- [ ] test should be changed according @congw729 please provide instructions for offline tests
- [ ] support image2video& txt-image2video jobs 
image
- [ ] refactor the examples/offline/video_generation/ which can be used for other video generation models

**Comment by congw729** [comment]
> I think we can get this PR merged now, later we need to open a new issue for a few todo jobs
> 
> * [ ]  test should be changed according @congw729 please provide instructions for offline tests
> * [ ]  support image2video& txt-image2video jobs
> 

Got it.

**Comment by pengchengneo** [comment]
Excellent work. 
May I ask a question, when implementing the text2video model, there should be some error accumulation due to precision transformation during the forward process. 
How do you perform model evaluation to ensure that the accuracy of our implemented model remains consistent with that in the paper? 
For example, text models can run on datasets like GSM8K to observe scores; how do we perform this kind of evaluation for video models? Thank you.
@linyueqian @hsliuustc0106
